### PR TITLE
Add dedicated e2e tester cmd for parallel conformance testing of kubermatic clusters

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
Adds a dedicated command to run multiple e2e tests in parallel.
It also adds retries in case a particular test fails.
There are 2 kinds of retries:
- Single test during conformance test suite fails -> retries are being done by Ginkgo
- Single test fails even after Ginkgo retries: command will restart the whole suite

Test results are being stored into a dedicated directory. Test result include:
- junit.xml
- logs of conformance tests

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: api/cmd/conformance-tests/main.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	glog.Info("Starting E2E tests...")
+
+	var scenarios []testScenario
+	scenarios = append(scenarios, getAWSScenarios()...)
+	scenarios = append(scenarios, getDigitaloceanScenarios()...)
+	scenarios = append(scenarios, getHetznerScenarios()...)
+	scenarios = append(scenarios, getOpenStackScenarios()...)
+
+	runner := newRunner(scenarios, &opts)
+	hadFailure, err := runner.Run()
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/main.go]: Why not return an error in case of failure?

**Comment by alvaroaleman** [comment]
Awesome pull, thanks for all the work!

One nit I would like to see changed longterm is that we don't have the providers specify the scenarios themselves but instead have them global and let the providers just specify why they don't run, so its opt-out rather than opt-in.

**Review by alvaroaleman — DISMISSED** [review]

File: api/cmd/conformance-tests/main.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	config, err := clientcmd.BuildConfigFromFlags("", opts.kubeconfigPath)
+	if err != nil {
+		glog.Fatal(err)
+	}
+
+	kubermaticClient := kubermaticclientset.NewForConfigOrDie(config)
+
+	if opts.cleanupOnStart {
+		if opts.namePrefix == "" {
+			log.Fatalf("cleanup-on-start was specified but name-prefix is empty")
```
Comment (p0lyn0mial) [inline api/cmd/conformance-tests/main.go]: replace `log.Fatal` with `glog.Fatal`

File: api/cmd/conformance-tests/main.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+				p := metav1.DeletePropagationBackground
+				opts := metav1.DeleteOptions{PropagationPolicy: &p}
+				fmt.Println("Deleting cluster " + cluster.Name + "...")
+				if err = kubermaticClient.KubermaticV1().Clusters().Delete(cluster.Name, &opts); err != nil {
+					log.Fatalf("failed to delete cluster %s: %v", cluster.Name, err)
+				}
+			}
+		}
+
+		fmt.Println("Cleaned up all old clusters")
```
Comment (p0lyn0mial) [inline api/cmd/conformance-tests/main.go]: `glog.Info` ?

**Review by p0lyn0mial — COMMENTED** [review]
Nice pull indeed, brings a huge value !

- Have you considered to write the tests as standard golang tests and use `go test` to run them ? (just asking)
- Do you have a test result I could see ?

**Comment by mrIncompetent** [comment]
@p0lyn0mial 

> Have you considered to write the tests as standard golang tests and use go test to run them ? (just asking)

Yes, i started using go tests first, but using a dedicated command felt simpler (code & usage) when using multiple flags. Thus i moved back to a simple command.

> Do you have a test result I could see ?

A test result from a scenario: https://storage.cloud.google.com/e2e-test-results/aws-ubuntu-1.10.8.zip

The folder structure is:
- reports (What you specify via flag)
  - scenario-name (For example: `aws-ubuntu-1.10.8`)
    - attempt-1
    - attempt-2 (In case a restart happened due to failures)
  - scenario-name (For example: `aws-ubuntu-1.11.3`)
    - attempt-1

**Comment by mrIncompetent** [comment]
@alvaroaleman @p0lyn0mial 
I fixed the issues, addresses the comments and rebased into master. 
PTAL

Adding more providers & making the tests in a opt-out instead opt-int (https://github.com/kubermatic/kubermatic/pull/2112#issuecomment-426174750) will be done in follow up PR's.

**Review by alvaroaleman — APPROVED** [review]

**Review by p0lyn0mial — APPROVED** [review]

**Pull request merged**